Demo bugfixes and improvements

roblim opened the pull request:
- fix kl_divergence bug in profiling results rendering to handle categorical distributions 
- add expect_column_quantile_values_to_be_between renderer
- add display of actual kl_divergence value to observed value column of validation results view (alongside graph of distribution)
- various ui refinements
- add content block container template
- create static assets directory and add ge fonts and images, copy dir tree during data_context scaffolding

roblim committed:
Copy data docs static assets directory

roblim committed:
Move logic into page action card

roblim committed:
Add static assets directory

roblim committed:
Add jinja global var for static_assets_dir

roblim committed:
Move styles into static dir

roblim committed:
Change headers to h4

roblim committed:
Add fonts to static dir

roblim committed:
Add logos and favicon to static images dir

roblim committed:
Add ge logo to top navbar; make logo link to home

roblim committed:
Add favicon

roblim committed:
Update syles dir

roblim committed:
Add favicon

roblim committed:
Nix dickens photo from ge_info; open ge link in new tab

roblim committed:
Add logic to handle categorical distributions

roblim committed:
Add expect_column_quantile_values_to_be_between renderer

roblim committed:
Add quantile strings for Q1, Q3 and Median; Add to validation results

roblim committed:
Add observed value to kl_divergence validation result chart

roblim committed:
Create content_block_container template

roblim committed:
Update styles template path

roblim committed:
Update tests

roblim committed:
Add data_context argument

roblim committed:
Update tests

roblim committed:
Update kl_divergence observed value display

- display "None (-infinity, infinity, or NaN)" when actual kl-divergence is None
- nix strong tag

roblim committed:
Fix typos

roblim committed:
Only copy static assets dir tree if doesn’t exist

pep8speaks commented:
Thank you for submitting to Great Expectations!



* In the file [`great_expectations/data_context/store/namespaced_read_write_store.py`](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/great_expectations/data_context/store/namespaced_read_write_store.py):

> [Line 335:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/great_expectations/data_context/store/namespaced_read_write_store.py#L335): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 343:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/great_expectations/data_context/store/namespaced_read_write_store.py#L343): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`great_expectations/data_context/store/store_backend.py`](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/great_expectations/data_context/store/store_backend.py):

> [Line 209:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/great_expectations/data_context/store/store_backend.py#L209): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace





* In the file [`great_expectations/render/renderer/content_block/validation_results_table_content_block.py`](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/great_expectations/render/renderer/content_block/validation_results_table_content_block.py):

> [Line 261:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/great_expectations/render/renderer/content_block/validation_results_table_content_block.py#L261): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`great_expectations/render/renderer/site_builder.py`](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/great_expectations/render/renderer/site_builder.py):

> [Line 107:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/great_expectations/render/renderer/site_builder.py#L107): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 203:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/great_expectations/render/renderer/site_builder.py#L203): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 206:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/great_expectations/render/renderer/site_builder.py#L206): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace









* In the file [`tests/store/test_store_backends.py`](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/store/test_store_backends.py):

> [Line 244:46](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/store/test_store_backends.py#L244): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 244:121](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/store/test_store_backends.py#L244): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (126 > 120 characters)
> [Line 269:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/store/test_store_backends.py#L269): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 271:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/store/test_store_backends.py#L271): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`tests/test_jupyter_ux.py`](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/test_jupyter_ux.py):

> [Line 47:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/test_jupyter_ux.py#L47): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 48:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/test_jupyter_ux.py#L48): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 54:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/test_jupyter_ux.py#L54): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 55:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/test_jupyter_ux.py#L55): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 157:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/test_jupyter_ux.py#L157): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 158:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/test_jupyter_ux.py#L158): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 165:1](https://github.com/great-expectations/great_expectations/blob/4e3af39e1465ad954d1dcad234f353d83c45d34a/tests/test_jupyter_ux.py#L165): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

This linter comment has been updated based on the most recent commit.

##### Comment last updated at 2019-12-04 22:05:06 UTC

roblim committed:
Add static assets dir to manifest

roblim committed:
Fix linter errors

jcampbell commented:
view appears to be using the data context root directory in this case, but that is not something that will be generally available across all site stores. Why can't the static assets directory be relative to the store base?

roblim committed:
Merge branch 'develop' into demo-bugfixes-and-improvements

roblim committed:
Nix scaffolding of static assets in init flow

- move to site builder

roblim committed:
Move static assets copying to site builder; set custom styles dir in site builder

- nix data_context init arg from View, passing in custom styles dir instead

roblim committed:
Use relative paths for asset dirs

roblim committed:
Update tests

roblim committed:
Fix merge conflicts

roblim committed:
Fix typo

roblim commented on great_expectations/render/renderer/site_builder.py:
```diff
@@ -100,6 +103,13 @@ def __init__(self,
+        self.store_backend = store_backend
+        
+        # set custom_styles_directory if present
+        custom_styles_directory = None
+        plugins_directory = data_context.plugins_directory
+        if os.path.isdir(os.path.join(plugins_directory, "custom_data_docs", "styles")):
+            custom_styles_directory = os.path.join(plugins_directory, "custom_data_docs/styles")
```
Moved definition of custom_styles_directory from View to SiteBuilder, which is passed into View __init__

roblim commented on great_expectations/render/renderer/site_builder.py:
```diff
@@ -188,7 +200,20 @@ def build(self, resource_identifiers=None):
+            logger.warning(message)
+        else:
+            shutil.copytree(static_assets_source_dir, static_assets_destination_path)
+        
```
Moved copying of static assets tree from ge init scaffolding to SiteBuilder build process

jcampbell replied:
Can we go over this in a call tomorrow? Since store_backend may not have a base_directory, I think site_base_dir still isn't being set correctly here.

roblim commented on great_expectations/render/view/templates/top_navbar.j2:
```diff
@@ -2,19 +2,32 @@
 {% endif %}
 
-<nav class="navbar bg-light navbar-expand-md sticky-top border-bottom">
+{% if utm_medium == "validation-results-page" or utm_medium == "profiling-results-page" %}
+  {% set home_url = "../../../../../index.html" -%}
+  {% set static_images_dir = "../../../../../static/images/" -%}
+{% elif utm_medium == "expectation-suite-page" %}
+  {% set home_url = "../../../../index.html" -%}
+  {% set static_images_dir = "../../../../static/images/" -%}
+{% elif utm_medium == "index-page" %}
+  {% set home_url = "#" -%}
+  {% set static_images_dir = "./static/images/" -%}
+{% endif %}
```
Switched to relative paths for static assets

roblim commented:
> view appears to be using the data context root directory in this case, but that is not something that will be generally available across all site stores. Why can't the static assets directory be relative to the store base?

@jcampbell - see latest changes to address above

jcampbell reviewed (changes requested):
Let's do a quick call to figure out the best option for where to root the static path

roblim committed:
Fix typo

roblim committed:
Use new copy_static_assets() store method to copy static assets

roblim committed:
Add new “static_assets” store backend

- special store with no key_length or filepath_template
- enables use of variable-length keys

roblim committed:
Create new copy_static_assets method

- uses .set() method of store backend to write files

roblim committed:
Add newline

roblim committed:
Allow key_length to be None

roblim committed:
Allow values to be of type bytes

roblim committed:
Allow variable-length keys when no filepath template is set

roblim committed:
Handle writing of bytes

roblim committed:
Merge branch 'develop' into demo-bugfixes-and-improvements

roblim committed:
Update test

roblim committed:
Add ability to skip dirs or files

jcampbell commented on great_expectations/data_context/store/store_backend.py:
```diff
@@ -191,16 +190,23 @@ def _convert_key_to_filepath(self, key):
 
-        # NOTE : These methods support fixed-length keys, but not variable.
+        # NOTE : These methods support variable-length keys if no filepath template is provided, in which case
+        # all key elements are joined to generate the filepath
```
+ When using variable-length keys, the backend cannot list keys

roblim replied:
Since static_assets are using special store with no key_length or filepath_template, list_keys just walks full_base_directory and does simple conversion since adding following to _convert_filepath_to_key:

```python
if not self.filepath_template:
    return tuple(filepath.split(os.sep))
```

jcampbell commented on great_expectations/data_context/store/store_backend.py:
```diff
@@ -191,16 +190,23 @@ def _convert_key_to_filepath(self, key):
+        if self.filepath_template:
+            converted_string = self.filepath_template.format(*list(key))
+        else:
+            converted_string = '/'.join(key)
```
Check whether we are supposed to use the os separator based on configuration here?

roblim replied:
This gets covered in the following lines, no?
```python
if self.platform_specific_separator:
    converted_string = os.path.join(*converted_string.split('/'))
return converted_string
```

jcampbell commented on great_expectations/data_context/store/store_backend.py:
```diff
@@ -387,7 +396,10 @@ def _set(self, key, value, content_encoding='utf-8', content_type='application/j
         s3 = boto3.resource('s3')
         result_s3 = s3.Object(self.bucket, s3_object_key)
-        result_s3.put(Body=value.encode(content_encoding), ContentEncoding=content_encoding, ContentType=content_type)
+        if type(value) is bytes:
```
1. whenvever checking type, we need to use isinstance(value, bytes) instead of type is, to allow subclasses to pass

roblim replied:
Ah, good point. Will change.

roblim committed:
Merge branch 'develop' into demo-bugfixes-and-improvements

jcampbell commented on great_expectations/data_context/store/namespaced_read_write_store.py:
```diff
@@ -320,3 +331,34 @@ def write_index_page(self, page):
+                    # Only use path elements starting from static/ for key
+                    store_key = tuple(os.path.normpath(source_name).split(os.sep))
+                    store_key = store_key[store_key.index('static'):]
+                    self.store_backends["static_assets"].set(store_key, f.read(), is_file=True)
```
I don't see is_file elsewhere...what is that controlling?

jcampbell commented on great_expectations/data_context/store/store_backend.py:
```diff
@@ -474,7 +486,10 @@ def _set(self, key, value, content_encoding='utf-8', content_type='application/j
         blob = bucket.blob(gcs_object_key)
-        blob.upload_from_string(value.encode(content_encoding), content_type=content_type)
+        if type(value) is bytes:
+            blob.upload_from_string(value)
```
I think we still want to set the content_type /encoding here.

jcampbell commented on great_expectations/data_context/store/store_backend.py:
```diff
@@ -391,7 +400,10 @@ def _set(self, key, value, content_encoding='utf-8', content_type='application/j
         result_s3 = s3.Object(self.bucket, s3_object_key)
-        result_s3.put(Body=value.encode(content_encoding), ContentEncoding=content_encoding, ContentType=content_type)
+        if type(value) is bytes:
+            result_s3.put(Body=value)
```
I think we still want to set content_type / content_encoding here

roblim replied:
K. Added logic to get content_type and content_encoding using mimetype.guess_type(), which is extension based. Tested it out with files that are currently in `static`, and get content_types just fine, but guess_type always yields None for encoding. Unfamiliar with content_encoding when it comes to non-text files. Since we're getting bytes using f.read() in binary mode, would encoding just be UTF-8?

jcampbell reviewed (changes requested):
Looks good; a few smallish updates

roblim committed:
Use isistance to check type

roblim committed:
Set content_encoding and content_type

roblim committed:
Get file content_type and encoding using mimetypes.guess_type

roblim committed:
Update test and add test case using backend with no filepath template

roblim committed:
Only pass content encoding for string types

roblim committed:
Update test

roblim committed:
Handle py2

- try/except is to support py2, since both str and bytes objects pass above condition

roblim committed:
Merge branch 'develop' into demo-bugfixes-and-improvements
